app/models/workflow.rb

```diff
@@ -1,34 +1,49 @@
 class Workflow
 
-  attr_accessor :steps
+  attr_accessor :first_step
 
   def self.build(definition)
     workflow = Workflow.new
-    workflow.steps = definition[:steps]
+
+    definition[:steps]
+      .each_with_index { |x, i| x[:next_step] = definition[:steps][i+1] }
+      .each_with_index { |x, i| x[:method] = lambda { |e| x[:type].constantize.new.receive e } }
+
+    workflow.first_step = definition[:steps].first
+    workflow.first_step[:method] = lambda { |e| workflow.first_step[:type].constantize.new.fire e }
+
     workflow
   end
 
   def start(data)
 
-    steps.each_with_index { |x, i| x[:method] = (i == 0 ? :fire : :receive) }
+    return if first_step.nil?
+
+    execute_step first_step, data
 
-    steps.reduce(data) do |last_event, step_data|
+  end
 
-      step = step_data[:type].constantize.new
+  private
 
-      event = step.send(step_data[:method], last_event)
+  def execute_step(step, event)
+
+    events = [step[:method].call(event)].flatten
+
+    events.each { |e| persist e, event }
+
+    return if step[:next_step].nil?
+
+    events.each { |e| execute_step step[:next_step], e }
```

Comment:
2 events were fired as a result of the last event, and we're sending both into the next step. I'm sure this will become a configuration choice in the future. it's this way now for simplification.